config/src/config/upstream_config.rs

```diff
@@ -1,52 +1,60 @@
 pub type UpstreamNetworkId = PeerId;
 
+/// If a node considers a network 'upstream', all peers of this node in that network are broadcast recipients
+/// for mempool, and sync-responders in state sync
+/// For validators, it is unnecessary to declare their validator network as their primary upstream network in this config
+/// Otherwise, any non-validator network not declared here will be treated as a downstream
+/// network (i.e. transactions will not be broadcast to and sync requests will not be sent to such networks)
 #[derive(Clone, Default, Debug, Deserialize, PartialEq, Serialize)]
 #[serde(default, deny_unknown_fields)]
 pub struct UpstreamConfig {
-    // primary upstream network ids. All peers in such network are used as upstream for this node
-    pub primary_networks: Vec<UpstreamNetworkId>,
-    // All upstream peers of this node, across all the networks that are statically defined in this node's config
-    // this is mostly meaningful in VFN networks, where there is a strict hierarchy in a network
-    pub upstream_peers: HashSet<PeerNetworkId>,
-    // optional fallback network id. Used to as a failover if preferred upstream peers are not available
-    // TODO replace PeerId with `NetworkConfig` to contain actual info needed to build fallback_network
-    pub fallback_networks: Vec<UpstreamNetworkId>,
+    // list of upstream networks for this node, ordered by preference
+    // A validator's primary upstream network is their validator network, and for a FN,
+    // it is the first network defined here. If the primary upstream network goes down, the node will fall back to the networks
+    // specified here, in this order
+    pub networks: Vec<NetworkId>,
 }
 
 impl UpstreamConfig {
     /// Determines whether a node `peer_id` in network `network_id` is an upstream peer of a node with this NodeConfig.
-    pub fn is_upstream_peer(&self, peer: PeerNetworkId) -> bool {
-        self.is_primary_upstream_peer(peer) || self.fallback_networks.contains(&peer.network_id())
+    pub fn is_upstream_peer(&self, peer: &PeerNetworkId) -> bool {
+        let network_id = peer.network_id();
+        network_id == NetworkId::Validator || self.networks.contains(&network_id)
     }
 
-    pub fn is_primary_upstream_peer(&self, peer: PeerNetworkId) -> bool {
-        self.primary_networks.contains(&peer.network_id()) || self.upstream_peers.contains(&peer)
+    pub fn is_primary_upstream_peer(&self, peer: &PeerNetworkId) -> bool {
+        let network_id = peer.network_id();
+        network_id == NetworkId::Validator || self.networks.get(0) == Some(&network_id)
     }
 }
 
-#[derive(Copy, Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
 /// Identifier of a node, represented as (network_id, peer_id)
-pub struct PeerNetworkId(pub UpstreamNetworkId, pub PeerId);
+pub struct PeerNetworkId(pub NetworkId, pub PeerId);
```

Comment:
feels like we don't need this any more?

Reply:
I was banging my head on this, and learned UpstreamNetworkId is not a NetworkId but is a PeerId and as a single node it can have several different peer ids in different networks. amazing

Reply:
so `PeerNetworkId` is used by mempool/state sync to identify a peer across all the networks it is involved in, since last time I checked there can be duplicate peer IDs across different networks..

Reply:
I see, I guess I had envisioned the model slightly differently, but I am giving feedback without digging into the code very well. I do reckon that some of this logic should probably be outside of the config space, but at the same time,  I think we should divide the needs and the wants.